genome_sampler/subsample_neighbors.py

```diff
@@ -0,0 +1,131 @@
+import tempfile
+
+import pandas as pd
+import numpy as np
+
+import qiime2
+from qiime2.metadata import CategoricalMetadataColumn
+from q2_types.feature_data import DNAFASTAFormat
+
+from genome_sampler.common import IDSelection, run_command, ids_from_fasta
+
+
+def _clusters_from_vsearch_out(vsearch_out, locale):
+    clusters = {}
+    if locale is None:
+        def f(context_id):
+            return None
+    else:
+        def f(context_id):
+            return locale[context_id]
+
+    for r in vsearch_out.itertuples():
+        if r.focal_id not in clusters:
+            clusters[r.focal_id] = []
+        clusters[r.focal_id].append((r.context_id,
+                                     r.n_mismatches,
+                                     f(r.context_id)))
+    columns = ['context_id', 'n_mismatches', 'locale']
+    return {focal_id: pd.DataFrame(hits, columns=columns)
+            for focal_id, hits in clusters.items()}
+
+
+def _sample_cluster(cluster, samples_per_cluster, random_state):
+    len_cluster = cluster.shape[0]
+    if len_cluster == 0:
+        return []
+
+    if len_cluster <= samples_per_cluster:
+        return list(cluster['context_id'])
+
+    unique_locales = cluster['locale'].unique()
+    if len(unique_locales) == 1:
+        # if only one locale, select context sequences spanning the range
+        # of number of mismatches
+        sorted_cluster = cluster.sort_values(by='n_mismatches')
+        indices = map(int, np.linspace(0, len_cluster-1, samples_per_cluster))
+        return [sorted_cluster.iloc[i]['context_id'] for i in indices]
+
+    # convert locations into number of obs, take the inverse, then scale by
+    # number of unique locations so it all sums to 1. Infrequent location obs
+    # will have a high weight, frequent location obs will have an individually
+    # lower weight, but collectively all locations have the same weight when
+    # all obs are summed within a location class.
+    loc = cluster['locale']
+    weights = ((1 / loc.map(loc.value_counts())) / len(loc.unique()))
+    # TODO: if multiple samples are being selected from the same locale, it
+    # would be good to sample across the number of mismatches to the focal
+    # sequence, as we're doing where there is only a single locale. To do that,
+    # I think we'd need to get the count of how many of each locale we want
+    # at random (like the rarefaction algorithm) and then use the approach
+    # above (which would then generalize to one or more locales).
+    result = cluster.sample(samples_per_cluster,
+                            weights=weights,
+                            random_state=random_state)
```

Comment:
this is missing the behavior:
 https://github.com/caporaso-lab/genome-sampler/blob/30a837b42f927d5d85c9edf01e9cbed3d28a9697/genome_sampler/sample-clusters.py#L116-L124

which I think is what the TODO is trying to get at, but I'm not sure how we would do that.